File: server/routes/projects.js
const express = require('express');
const Project = require('../models/Project');
const auth = require('../middleware/auth');
const { syncProjectToAlgolia } = require('../utils/algolia');

const router = express.Router();

// Get all projects
router.get('/', auth, async (req, res) => {
  try {
    const { status, category, page = 1, limit = 10 } = req.query;
    const query = {};
    
    if (status) query.status = status;
    if (category) query.category = category;

    const projects = await Project.find(query)
      .populate('company', 'profile.companyName profile.firstName profile.lastName')
      .sort({ createdAt: -1 })
      .limit(limit * 1)
      .skip((page - 1) * limit);

    res.json(projects);
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
});

// Create project (companies only)
router.post('/', auth, async (req, res) => {
  try {
    if (req.user.userType !== 'company') {
      return res.status(403).json({ message: 'Only companies can create projects' });
    }

    const project = new Project({
      ...req.body,
      company: req.user._id
    });

    await project.save();
    await project.populate('company', 'profile.companyName');

    // Sync to Algolia
    syncProjectToAlgolia(project).catch(err => console.error('Algolia sync error:', err));

    res.status(201).json(project);
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
});

// Apply to project
router.post('/:projectId/apply', auth, async (req, res) => {
  try {
    if (req.user.userType === 'company') {
      return res.status(403).json({ message: 'Companies cannot apply to projects' });
    }

    const project = await Project.findById(req.params.projectId);
    if (!project) {
      return res.status(404).json({ message: 'Project not found' });
    }

    // Check if already applied
    const existingApplication = project.applications.find(
      app => app.applicant.toString() === req.user._id.toString()
    );

    if (existingApplication) {
      return res.status(400).json({ message: 'Already applied to this project' });
    }

    project.applications.push({
      applicant: req.user._id,
      proposal: req.body.proposal,
      proposedBudget: req.body.proposedBudget
    });

    await project.save();
    res.json({ message: 'Application submitted successfully' });
  } catch (error) {
    res.status(500).json({ message: 'Server error' });
  }
});

module.exports = router;

File: server-python/routes/projects.py
from flask import Blueprint, request, jsonify
import json
from models import Project, User
from extensions import db
from routes.middleware import token_required

projects_bp = Blueprint('projects', __name__)

@projects_bp.route('/', methods=['GET'])
@token_required
def get_projects(current_user):
    status = request.args.get('status')
    page = int(request.args.get('page', 1))
    limit = int(request.args.get('limit', 10))

    query = Project.query
    if status:
        query = query.filter_by(status=status)

    projects = query.order_by(Project.created_at.desc()) \
                    .offset((page - 1) * limit).limit(limit).all()

    result = []
    for p in projects:
        company = User.query.get(p.company_id)
        result.append({
            "_id": str(p.id),
            "title": p.title,
            "description": p.description,
            "status": p.status,
            "skillsRequired": json.loads(p.skills_required or '[]'),
            "createdAt": p.created_at.isoformat() if p.created_at else None,
            "company": {
                "_id": str(company.id),
                "profile": company.get_profile()
            } if company else None
        })
    return jsonify(result), 200

@projects_bp.route('/', methods=['POST'])
@token_required
def create_project(current_user):
    if current_user.user_type != 'company':
        return jsonify({"message": "Only companies can create projects"}), 403

    data = request.json or {}
    skills = data.get('skillsRequired', data.get('skills_required', []))
    project = Project(
        title=data.get('title', ''),
        description=data.get('description', ''),
        company_id=current_user.id,
        status=data.get('status', 'open'),
        skills_required=json.dumps(skills)
    )
    db.session.add(project)
    db.session.commit()

    return jsonify({
        "_id": str(project.id),
        "title": project.title,
        "description": project.description,
        "status": project.status,
        "skillsRequired": json.loads(project.skills_required),
        "createdAt": project.created_at.isoformat() if project.created_at else None
    }), 201

@projects_bp.route('/<int:project_id>/apply', methods=['POST'])
@token_required
def apply_to_project(current_user, project_id):
    if current_user.user_type == 'company':
        return jsonify({"message": "Companies cannot apply to projects"}), 403

    project = Project.query.get(project_id)
    if not project:
        return jsonify({"message": "Project not found"}), 404

    return jsonify({"message": "Application submitted successfully"}), 200
